numpy/core/shape_base.py

```diff
@@ -423,32 +437,90 @@ def format_index(index):
     """
-    def atleast_nd(a, ndim):
-        # Ensures `a` has at least `ndim` dimensions by prepending
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    for shape in shapes[1:]:
+        if (shape[:axis] != first_shape[:axis] or
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
 
-    return block_recursion(arrays)
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
+
+    return tuple(accumulated)
+
+
+def _block_info_recursion(arrays, list_ndim, result_ndim, depth=0):
+    """Form required information to block the arrays using a flat iterator.
+
+    Returns
+    -------
+    shape: tuple of ints
+        the resulting shape of the final array
+    region_slices: iterable of indices
+        slices into the full array required for assignment
+    region_data: iterable of ndarray
+        the data to assign to each slice of the full array
+
+    """
+    if depth < list_ndim:
+        axis = result_ndim - list_ndim + depth
+        info = [_block_info_recursion(arr, list_ndim, result_ndim, depth + 1)
+                for arr in arrays]
+        shapes, slices, arrays = zip(*info)
+        # Compute the offset for each array
+        offsets = (0,) + tuple(_accumulate(shape[axis]
+                                           for shape in shapes[:-1]))
+        # `slices` and `arrays` contain lists that have the information
+        # from the inner lists provided to the concatenation function.
+        # To create the correct offset, one needs to prepend the appropriate
+        # offset to each of them. Each list (containing one or more arrays)
+        # will require the matching offset (computed above with accumulate)
+        slice_prefixes = [(slice(offset, offset + shape[axis]),)
+                          for offset, shape in zip(offsets, shapes)]
```

Comment:
This assignment, and the one above, could be extracted to a helper function too:
```
def _concatenate_slices(shapes, axis):
    """
    Produce the slices that match the destination each array would end up along axis.

    For the following situation::

        ret = concatenate([a, b, c], axis=0)
        sl_a, sl_b, sl_c = concatenate_slices([a, b, c], axis=0)

    The result will be that `ret[sl_a] == a`, `ret[sl_b] == b`, and `ret[sl_c] == c`. Similar
    properties hold for other axes.
    """
    offsets = (0,) + tuple(_accumulate(shape[axis]
                                           for shape in shapes[:-1]))
    return [
         slice(offset, offset + shape[axis])
         for offset, shape in zip(offsets, shapes)
    ]
```

Reply:
I guess I don't really see the point of having so many helper functions. If somebody wants to extract this code, they can find the lines they need no?

Reply:
Furthermore, you may as well drop the `itertools.chain` and do it yourself:
```
slice_prefixes = _concatenate_slices(shapes, axis)
slices = [
    (slice_prefix,) + the_slice
    for slice_prefix, inner_slices in zip(slice_prefixes, slices)
    for the_slice in inner_slices
]
```

Reply:
done

Reply:
> I don't really see the point of having so many helper functions. 

It's not just about code reuse - it's about using function names to describe pieces of your algorithm instead of comments, and allowing the reader to reason about your code in pieces rather than all at once.

Reply:
got it. thanks.